examples/todoapp/lib/base/repositories/todo_repository.dart

```diff
@@ -1,45 +1,204 @@
+        () async {
+          final result = await dataSource
+              .updateCompletedById(id, {'completed': completed});
+          localDataSource.updateCompletedById(
+            id,
+            completed,
+          );
+          return result;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.updateCompletedById(
+            id,
+            completed,
+            synced: false,
+            action: TodoModelActions.update.name,
+          ),
+        ),
       );
 
-  Future<List<TodoModel>> updateCompletedForAll(bool completed) =>
+  Future<List<$TodoModel>> updateCompletedForAll(bool completed) =>
       _errorMapper.execute(
-        () => dataSource.updateCompletedForAll({'completed': completed}),
+        () async {
+          final result =
+              await dataSource.updateCompletedForAll({'completed': completed});
+          localDataSource.updateCompletedForAll(completed);
+          return result;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.updateCompletedForAll(
+            completed,
+            synced: false,
+            action: TodoModelActions.update.name,
+          ),
+        ),
       );
 
-  Future<List<TodoModel>> deleteCompleted() => _errorMapper.execute(
-        () => dataSource.deleteCompleted(),
+  Future<List<$TodoModel>> deleteCompleted() => _errorMapper.execute(
+        () async {
+          final response = await dataSource.deleteCompleted();
+          localDataSource.deleteCompleted();
+          return response;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.softDeleteCompleted(
+            synced: false,
+            action: TodoModelActions.delete.name,
+          ),
+        ),
       );
 
   Future<void> deleteTodoById(String id) => _errorMapper.execute(
-        () => dataSource.deleteTodoById(id),
+        () async {
+          await dataSource.deleteTodoById(id);
+          localDataSource.deleteTodoById(id);
+        },
+      ).onError(
+        (error, stackTrace) {
+          handleError(
+            error,
+            localDataSource.softDeleteTodoById(
+              id,
+              synced: false,
+              action: TodoModelActions.delete.name,
+            ),
+          );
+        },
       );
 
-  Future<TodoModel> fetchTodoById(String id) => _errorMapper.execute(
-        () => dataSource.getTodoById(id),
+  Future<$TodoModel> fetchTodoById(String id) => _errorMapper.execute(
+        () async {
+          final result = await dataSource.getTodoById(id);
+          return result;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.getTodoById(id),
+        ),
       );
+
+  Future<void> syncronize() async {
```

Comment:
Considering the sync logic might grow in terms of complexity let's move it to the business layer and please add unit tests

Reply:
Moved to service